ENH: nan handling in functions that use `stats._chk_asarray`

aeklant opened the pull request:
closes gh-571

aeklant commented:
A few implementation details:

On `_contains_nan`:
* This is a new helper function.
* It is just a one liner, but I put it in anyway because of two reasons:
  * maintainability: changing the implementation here is easier than changing it in all of the functions.
  * clarity: it looks tidier than calling `np.isnan(np.sum(x))`.
* The reason why it calls `np.sum` is to avoid creating a huge array into memory (e.g. `np.isnan(x).any`). I also thought of doing it with `np.dot`; it seems to run faster but the risk of overflowing is greater (or so I read).

On `mode`:
* There is a special case when there are strings in the array because `np.isnan` does not work on those so I handled it separately.

On everything else:
* I found issues running the regression tests when delegating to `mstats`, there seems to be a problem when trying to import from `mstats` into `stats`. I'm not sure what was happening so if we decide for that approach instead of the one I took here I would have to research it further.
* I expect the rest of the functions to look very similar to what I did here, so I wanted to ping you to reach a consensus before I go all out with the rest (to avoid rework).

Please feel free to give out your comments.

ev-br commented on scipy/stats/stats.py:
```diff
@@ -237,6 +237,10 @@ def _chk2_asarray(a, b, axis):
 
 
+def _contains_nan(a):
+    return np.isnan(np.sum(a))
```
It's worth a comment here as to why not ` isnan(a).any()` (what you wrote in the GH comment)

ev-br commented on scipy/stats/stats.py:
```diff
@@ -640,10 +648,29 @@ def mode(a, axis=0):
+    except TypeError:
+        contains_nan = False
+        warnings.warn("the input array cannot be properly checked for nan "
+                      "values. nan values will be ignored.", RuntimeWarning)
```
I'd put this loigic into the nan-check helper.

rgommers replied:
I agree, put as much in the helper function as possible.

ev-br commented:
What are the issues with using mstats?

aeklant commented:
@ev-br I didn't yet look very deep into it but I'm almost certain the problem is that importing `mstats` causes a circular import.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -640,10 +648,29 @@ def mode(a, axis=0):
+        return ModeResult(np.nan, np.nan)
+
+    elif contains_nan and nan_check == 'raise':
+        raise ValueError("The input contains nan values.")
```
This can also go inside the helper function.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -850,16 +881,27 @@ def tmin(a, lowerlimit=None, axis=0, inclusive=True):
     """
+    if nan_check not in ('undefined', 'raise', 'ignore'):
+        raise ValueError("nan_check must be either 'undefined', 'raise', or"
+                         " 'ignore'")
```
This can also go inside the helper function.

aeklant commented:
Notes:
* Most of the functionality was moved to the helper function. It is now in charge of raising the exceptions while each function will handle the `omit` and `propagate` cases.
* `nan_check` was changed to `nan_policy` because it sounds clearer than `nan_check` or `nan_action` (to me). This is of course arbitrary and up for discussion.
* `undefined` was changed to `propagate` and `ignore` was changed to `omit` as suggested in the discussion.
* I changed the expected value (in the regression tests) of the `ttest_*` functions when there is nan input. Please review to make sure this is correct and won't cause issues (like breaking backwards compat).
* IMPORTANT: the implementation of the `omit` policy worked on `mode`, `tmin` and `tmax` because it was already implemented (perhaps wrongly so) i.e. it didn't delegate to `mstats` but reimplemented the functionality with masked arrays. The rest of the functions should delegate to `mstats` so we will need to come up with a reasonable strategy to resolve the import conflict in order to be able to use those functions without having to re-implement and without breaking code.

aeklant commented on scipy/stats/stats.py:
```diff
@@ -601,7 +623,7 @@ def hmean(a, axis=0, dtype=None):
 
 
-def mode(a, axis=0):
+def mode(a, axis=0, nan_policy='propagate'):
```
`mode`'s `nan_policy` default should probably be 'omit' for backwards compatibility.

rgommers replied:
Why is that backwards compatible? ``'propagate'`` should do the right thing here, if you let it actually propagate (note that that doesn't mean "return nans" in this case).

rgommers replied:
If you just remove the special casing of ``nan_policy == 'propagate':``, then you should get the following:

   >>> stats.mode([np.nan, np.nan, 3])
   nan
   >>> stats.mode([np.nan, 3, 3])
   3

rgommers replied:
@aeklant ping, this comment still needs addressing.

rgommers replied:
Addressed now.

aeklant commented on scipy/stats/stats.py:
```diff
@@ -1039,6 +1095,16 @@ def moment(a, moment=1, axis=0):
+
+    # TODO: implement the delegation to mstats when nan_policy == 'omit'
+    # if contains_nan and nan_policy == 'omit':
+    #     a = ma.masked_invalid(a)
```
This is commented until we come up with a way to deal with the problem importing `mstats`.

rgommers replied:
From ``$ grin "= stats." mstats_basic.py`` you can see that ``stats`` is hardly used within ``mstats`` except for copying some docstrings. So the circular imports can be broken like so I think:

- create a new file ``mstats_add_docs.py`` which imports ``mstats_basic`` and ``stats``
- import ``mstats_add_docs`` in ``mstats``; don't import ``mstats_basic`` there
- import ``mstats_basic`` in ``stats.py`` and use it to forward the ``nan_policy='omit'`` cases.
- remove ``from . import stats`` from ``mstats_basic.py``

aeklant commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2419,7 +2539,8 @@ def test_ttest_1samp_new():
         anan = np.array([[1,np.nan],[-1,1]])
-        assert_equal(stats.ttest_1samp(anan, 0),([0, np.nan], [1,np.nan]))
+        # assert_equal(stats.ttest_1samp(anan, 0),([0, np.nan], [1,np.nan]))
+        assert_equal(stats.ttest_1samp(anan, 0),(np.nan, np.nan))
```
nan input test was modified. Please review to make sure it won't cause problems. Like backwards compat break.

ev-br commented:
This needs a rebase. I wonder if a rebase is going to poke Travis CI, too.

aeklant commented:
Rebase done.

rgommers commented:
I like the omit/propagate naming, quite a bit clearer.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -237,6 +237,28 @@ def _chk2_asarray(a, b, axis):
+    try:
+        # Calling np.sum to avoid creating a huge array into memory
+        # e.g. np.isnan(a).any()
+        contains_nan = np.isnan(np.sum(a))
```
Noticing now that ``np.sum(a)`` will emit a warning for invalid values, which is probably undesired here. So might want to wrap this inside

    with np.errstate(invalid='ignore'):

rgommers commented on scipy/stats/stats.py:
```diff
@@ -237,6 +237,28 @@ def _chk2_asarray(a, b, axis):
+        contains_nan = np.isnan(np.sum(a))
+    except TypeError:
+        # If the check cannot be properly performed we fallback to omiting
+        # nan values and raising a warning.
```
The only thing that now triggers this is ``mode`` with an array of strings right? Might want to mention that, otherwise it's a bit unclear where the ``TypeError`` could come from.

rgommers commented:
@aeklant I took the parts from your WIP branch that looked good and separated them out in 1 commit per file, then finished the rest to break the circular imports: https://github.com/rgommers/scipy/tree/nan_policy

ev-br commented on scipy/stats/stats.py:
```diff
@@ -237,6 +237,28 @@ def _chk2_asarray(a, b, axis):
+def _contains_nan(a, nan_policy='propagate'):
+    if nan_policy not in ('propagate', 'raise', 'omit'):
+        raise ValueError("nan_policy must be either 'propagate', 'raise', or "
+                         "'ignore'")
```
Typo: is it ignore or omit

aeklant commented:
@rgommers it looks like I may have done something wrong with the procedure. 

I created a new branch with the changes you added to mine (on your own branch), then made a few more changes, and pushed to this same open PR. After that I rebased because it showed some merge conflicts and pushed again. Could I get some tech assistance? Would it be better to just close this PR and open a new one?

rgommers commented:
I see no new commits. Right now there are 11 commits here - first 5 of yours that are older, then 5 of mine, and then a merge commit. Does the merge commit contain your new code? If so I would do the following (make a backup of your work for safety first):

    $ git reset c79b21e95f   # undo the merge commit, keep other commits
    $ # Now re-commit changes you want one by one
    $ git push origin _chk_asarray -f  # need to force-push, because of the merge commit that disappeared

rgommers commented:
Opening a new PR doesn't help usually. If you can fix your branch locally (which is needed either way), then you can push it here as easily as opening a new PR.

aeklant commented:
A few notes:
* I added the `equal_var` keyword to `mstats.ttest_ind`. Please review for correctness.
* I added the `bias` keyword to `mstats.describe` just to make it consistent to `stats.describe` but it is not implemented (in neither of them). This may be outside this issue's scope but it's worth noting.
* All the expected results for the `omit` keyword test cases were obtained by removing the values that would be set to nan and running them with the default nan policy (so it runs with `stats` instead of being delegated to `mstats`).
* In the new `_stats_mstats_common` module, the examples are being run as if they were test cases, causing the Travis CI to fail due to things like an extra white-space in the documentation. Is this normal and expected? and if so, how could I replicate it in a local run test? it gets annoying to have to amend a commit that has already been pushed just for a typo over and over again.

rgommers commented:
Regarding tests of examples, those are doctests which are run in the newly added "refguide check" build on TravisCI. You can run those with the ``--refguide-check`` argument to ``runtests.py``. It prevents us from merging examples that are either completely broken or have actual output different from what's displayed in the docstring.

rgommers commented:
> I added the bias keyword to mstats.describe just to make it consistent to stats.describe but it is not implemented (in neither of them). This may be outside this issue's scope but it's worth noting.

It's trivial to implement, looks like all that is needed is to add ``bias=bias`` to the calls to ``skew`` and ``kurtosis``. So just add it to this PR I'd say - less work than opening an issue and fixing in a separate PR.

rgommers commented:
> All the expected results for the omit keyword test cases were obtained by removing the values that would be set to nan and running them with the default nan policy (so it runs with stats instead of being delegated to mstats).

That's a good approach I think.

rgommers commented:
> I added the equal_var keyword to mstats.ttest_ind. Please review for correctness.

The changes look correct, however the ``equal_var`` keyword in mstats is not tested. Can be added simply by testing against the nonmasked version (add in ``TestTtest_ind.test_vs_nonmasked``).

rgommers commented:
So in summary, this PR is now almost ready but there are 3 minor things to address still:
- test for ``equal_var`` in ``mstats.ttest_ind``
- implement ``bias`` keyword in ``stats.describe`` and ``mstats.describe``
- address the comment on ``'propagate'`` in ``stats.mode``

rgommers commented:
@ev-br @josef-pkt since the refactoring that yielded ``_stats_mstats_common.py`` was mine, it may be good if one of you has a quick look at it before I merge this PR.

ev-br commented:
@rgommers  it looks quite good to me. I agree with your comment above on three minor things to address before merging.

rgommers commented:
@ev-br thanks

aeklant commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2229,7 +2333,7 @@ def test_ttest_rel():
         anan = np.array([[1, np.nan], [-1, 1]])
         assert_equal(stats.ttest_rel(anan, np.zeros((2, 2))),
-                     ([0, np.nan], [1, np.nan]))
+                     (np.nan, np.nan))
```
This hasn't been addressed. I think it is an appropriate change but it does change the expected output so it has potential to break some code. Do you think this is something we should consider at least mentioning somewhere? or is it ok to just go ahead with it?

rgommers replied:
Good catch. The change isn't right. This demonstrates that the behavior for nan propagation currently in master works fine as is:

    In [1]: from scipy import stats

    In [2]: a2 = stats.norm.rvs(size=9).reshape(3,3)

    In [3]: b2 = np.arange(9).reshape(3,3)

    In [4]: stats.ttest_rel(a2, b2)
    Out[4]: Ttest_relResult(statistic=array([-1.44542944, -1.70106168, -1.93403464]), pvalue=array([ 0.28521702,  0.23103717,  0.19278399]))

    In [5]: a2[0,0] = np.nan

    In [6]: stats.ttest_rel(a2, b2)   # default is axis=0, so only expect a nan for the first axis here
    Out[6]: Ttest_relResult(statistic=array([        nan, -1.70106168, -1.93403464]), pvalue=array([        nan,  0.23103717,  0.19278399]))

That's why I just commented about the 2 lines to be removed from ``ttest_rel``. And so the change to the test here needs to be undone.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -3880,6 +3881,20 @@ def ttest_rel(a, b, axis=0):
+        return mstats_basic.ttest_rel(a, b, axis)
+
+    if contains_nan and nan_policy == 'propagate':
+        return Ttest_relResult(np.nan, np.nan)
```
@aeklant these 2 lines need to be deleted (same reasoning as we just discussed for ``stats.mode``).

aeklant commented:
Remaining action items have been addressed (and also the `ttest*` functions default behaviour). Please review.

rgommers commented:
Next request for an ``'omit'`` option (in ``stats.median_test``) in gh-5114. Once this PR is merged (not now) we should make an overview of all other functions for which the new kw could/should be added.

rgommers commented:
Reviewed the last commits, all look OK now. Except for finalizing the ``nan_policy`` name (or changing it), this is ready to merge.

aeklant commented:
I added `nan_policy` to `stats.kruskal`. Please refer to gh-5126 for details.

rgommers commented:
Last commit looks fine, except for the nitpick that ``kargs`` should be ``kwargs``. I'll merge this and fix that after. This PR has been open for long enough......

rgommers commented:
Thanks a lot @aeklant and everyone who contributed. Nice new feature!
